File: app/src/main/java/it/gruppoinfor/home2work/api/SessionManager.java
package it.gruppoinfor.home2work.api;

import android.content.Context;
import android.content.SharedPreferences;

import it.gruppoinfor.home2work.api.Client;
import it.gruppoinfor.home2work.models.User;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * Created by Federico on 10/01/2017.
 * Permette di salvare la sessione dell'utente nelle preferenze per un login rapido
 */

public class SessionManager {

    public static final String AUTH_CODE = "auth_code";
    private final SharedPreferences prefs;
    private final Context context;
    private final String IS_LOGIN = "IsLoggedIn";
    private final String KEY_TOKEN = "token";
    private final String KEY_EMAIL = "email";
    private final String KEY_ID = "id";

    public SessionManager(Context context) {
        this.context = context;
        prefs = context.getSharedPreferences("it.fleetup.app.session", Context.MODE_PRIVATE);
    }

    public void storeSession(final User signedUser) {

        // Salva le informazioni in modo persistente
        SharedPreferences.Editor editor = prefs.edit();
        editor.putBoolean(IS_LOGIN, true);
        editor.putString(KEY_EMAIL, signedUser.getEmail());
        editor.putString(KEY_TOKEN, signedUser.getToken());
        editor.putLong(KEY_ID, signedUser.getId());
        editor.apply();

        // TODO Controlla il token Firebase Cloud Messaging
        /* String fcmToken = FirebaseInstanceId.getInstance().getToken();

        if (!account.getFcmToken().equals(fcmToken)) {
            account.setFcmToken(fcmToken);
            Client.getAPI().updateUser(account).enqueue(new retrofit2.SessionManagerCallback<Account>() {
                @Override
                public void onResponse(Call<Account> call, Response<Account> response) {

                }

                @Override
                public void onFailure(Call<Account> call, Throwable t) {

                }
            });
        }*/
    }

    public User loadSession() {
        if (!isUserSignedIn())
            return null;
        User user = new User();
        user.setId(prefs.getLong(KEY_ID, 0));
        user.setEmail(prefs.getString(KEY_EMAIL, null));
        user.setToken(prefs.getString(KEY_TOKEN, null));
        return user;
    }

    public void checkSession(final SessionManagerCallback callback) {

        if (!isUserSignedIn()) {
            callback.onInvalidSession(AuthCode.NO_SESSION);
        } else {

            String email = prefs.getString(KEY_EMAIL, "");
            String password = prefs.getString(KEY_TOKEN, "");

            Client.getAPI().login(email, password).enqueue(new Callback<User>() {
                @Override
                public void onResponse(retrofit2.Call<User> call, Response<User> response) {
                    switch (response.code()) {
                        case 200:
                            User user = response.body();
                            storeSession(user);
                            callback.onValidSession(user);
                            break;
                        default:
                            callback.onInvalidSession(AuthCode.EXPIRED_TOKEN);
                            break;
                    }
                }

                @Override
                public void onFailure(retrofit2.Call<User> call, Throwable t) {
                    callback.onInvalidSession(AuthCode.EXPIRED_TOKEN);
                }
            });
        }
    }

    public void signOutUser() {

        // Elimina tutte le preferenze (com.infor.fleetupht.session)
        SharedPreferences.Editor editor = prefs.edit();
        editor.clear();
        editor.apply();

        // Termina i servizi
        // TODO context.stopService(new Intent(context, RouteService.class));

    }

    public boolean isUserSignedIn() {
        return prefs.getBoolean(IS_LOGIN, false);
    }

    public enum AuthCode {
        EXPIRED_TOKEN, SIGNED_OUT, NO_SESSION
    }

    public interface SessionManagerCallback {
        void onValidSession(User account);

        void onInvalidSession(AuthCode code);
    }

}

File: App/src/main/java/it/gruppoinfor/home2work/auth/SessionManager.kt
package it.gruppoinfor.home2work.auth

import android.annotation.SuppressLint
import android.content.Context
import android.content.Intent
import com.crashlytics.android.Crashlytics
import com.crashlytics.android.answers.Answers
import com.crashlytics.android.answers.LoginEvent
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import it.gruppoinfor.home2work.api.HomeToWorkClient
import it.gruppoinfor.home2work.api.RetrofitException
import it.gruppoinfor.home2work.tracking.LocationService
import it.gruppoinfor.home2work.tracking.SyncJobService
import it.gruppoinfor.home2work.user.User


/**
 * Created by Federico on 10/01/2017.
 * Permette di salvare la sessione dell'utente nelle preferenze e di ripristinarla all'occorrenza (se non scaduta)
 */

object SessionManager {

    private const val PREFS_SESSION = "it.home2work.app.session"
    private const val PREFS_TOKEN = "ACCESS_TOKEN"

    fun storeSession(ctx: Context, user: AuthUser?) {
        if (user == null) return

        val prefs = ctx.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE)
        val editor = prefs.edit()

        editor.putString(PREFS_TOKEN, user.accessToken)
        editor.apply()

    }

    @SuppressLint("CheckResult")
    fun loadSession(context: Context, callback: SessionCallback) {
        val prefs = context.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE)

        val user: User? = HomeToWorkClient.user

        if (user != null) {
            callback.onValidSession()
            return
        }

        val token = prefs.getString(PREFS_TOKEN, "")

        if (token.isEmpty()) {
            callback.onInvalidSession(SignInActivity.CODE_NO_ACCESS_TOKEN, null)
            return
        }

        HomeToWorkClient.getAuthService().login(token)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .onErrorReturn { null }
                .subscribe({

                    HomeToWorkClient.user = it

                    // Crashlytics log
                    Answers.getInstance().logLogin(LoginEvent()
                            .putMethod("Token")
                            .putSuccess(true))

                    Crashlytics.setUserIdentifier(HomeToWorkClient.user?.id.toString())
                    Crashlytics.setUserEmail(HomeToWorkClient.user?.email)
                    Crashlytics.setUserName(HomeToWorkClient.user.toString())

                    callback.onValidSession()

                }, {

                    clearSession(context)
                    Answers.getInstance().logLogin(LoginEvent()
                            .putMethod("Token")
                            .putSuccess(false))

                    when ((it as RetrofitException).kind) {
                        RetrofitException.Kind.NETWORK -> {
                            callback.onInvalidSession(SignInActivity.CODE_LOGIN_ERROR, null)
                        }
                        RetrofitException.Kind.HTTP -> {
                            callback.onInvalidSession(SignInActivity.CODE_INVALID_CREDENTIALS, null)
                        }
                        RetrofitException.Kind.UNEXPECTED -> {
                            callback.onInvalidSession(SignInActivity.CODE_SERVER_ERROR, it)
                        }
                    }
                })

    }


    fun clearSession(context: Context) {
        context.stopService(Intent(context, LocationService::class.java))
        SyncJobService.remove(context)

        val prefs = context.getSharedPreferences(PREFS_SESSION, Context.MODE_PRIVATE)
        val editor = prefs.edit()
        editor.clear()
        editor.apply()
    }

    interface SessionCallback {
        fun onValidSession()
        fun onInvalidSession(code: Int, throwable: Throwable?)
    }

}
